modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/leases/LeaseTracker.java

```diff
@@ -257,4 +252,57 @@ private PendingIndependentComparableValuesTracker<HybridTimestamp, ReplicaMeta>
     ) {
         return primaryReplicaWaiters.computeIfAbsent(groupId, key -> new PendingIndependentComparableValuesTracker<>(MIN_VALUE));
     }
+
+    private CompletableFuture<Void> loadLeasesBusyAsync(long recoveryRevision) {
+        Entry entry = msManager.getLocally(PLACEMENTDRIVER_LEASES_KEY, recoveryRevision);
+
+        CompletableFuture<Void> loadLeasesFuture;
+
+        if (entry.empty() || entry.tombstone()) {
+            leases = new Leases(Map.of(), BYTE_EMPTY_ARRAY);
+
+            loadLeasesFuture = completedFuture(null);
+        } else {
+            byte[] leasesBytes = entry.value();
+
+            LeaseBatch leaseBatch = LeaseBatch.fromBytes(ByteBuffer.wrap(leasesBytes).order(LITTLE_ENDIAN));
+
+            Map<ReplicationGroupId, Lease> leasesMap = new HashMap<>();
+
+            List<CompletableFuture<?>> fireEventFutures = new ArrayList<>();
+
+            leaseBatch.leases().forEach(lease -> {
+                leasesMap.put(lease.replicationGroupId(), lease);
+
+                if (lease.isAccepted()) {
+                    getOrCreatePrimaryReplicaWaiter(lease.replicationGroupId()).update(lease.getExpirationTime(), lease);
+
+                    fireEventFutures.add(fireEventReplicaBecomePrimary(recoveryRevision, lease));
+                }
+            });
+
+            leases = new Leases(unmodifiableMap(leasesMap), leasesBytes);
+
+            loadLeasesFuture = allOf(fireEventFutures.toArray(CompletableFuture[]::new));
+        }
+
+        LOG.info("Leases cache recovered [leases={}]", leases);
+
+        return loadLeasesFuture;
+    }
+
+    private CompletableFuture<Void> fireEventReplicaBecomePrimary(long causalityToken, Lease lease) {
+        String leaseholder = lease.getLeaseholder();
+
+        assert leaseholder != null : lease;
+
+        return fireEvent(
+                PRIMARY_REPLICA_ELECTED,
+                new PrimaryReplicaEventParameters(causalityToken, lease.replicationGroupId(), leaseholder)
+        );
+    }
+
+    private static boolean needFireEventReplicaBecomePrimary(@Nullable Lease previousLease, Lease newLease) {
```

Comment:
Please add a comment explaining what's going on here. I'd also add an assertion that newLease is accepted.

Reply:
Fix it